pkg/webhook/aws_secret_manager.go

```diff
@@ -0,0 +1,152 @@
+// Otherwise, the Pod will fail with an init-error.
+// Files will be mounted on /etc/flyte/secrets/<SecretGroup>/<SecretKey>
+type AWSSecretManagerInjector struct {
+	cfg config.AWSSecretManagerConfig
+}
+
+func formatAWSSecretArn(secret *core.Secret) string {
+	return strings.TrimRight(secret.Group, ":") + ":" + strings.TrimLeft(secret.Key, ":")
+}
+
+func formatAWSInitContainerName(index int) string {
+	return fmt.Sprintf("aws-pull-secret-%v", index)
+}
+
+func (i AWSSecretManagerInjector) Type() config.SecretManagerType {
+	return config.SecretManagerTypeAWS
+}
+
+func (i AWSSecretManagerInjector) Inject(ctx context.Context, secret *core.Secret, p *corev1.Pod) (newP *corev1.Pod, injected bool, err error) {
+	if len(secret.Group) == 0 || len(secret.Key) == 0 {
+		return nil, false, fmt.Errorf("AWS Secrets Webhook require both key and group to be set. "+
+			"Secret: [%v]", secret)
+	}
+
+	switch secret.MountRequirement {
+	case core.Secret_ANY:
+		fallthrough
+	case core.Secret_FILE:
+		// A Volume with a static name so that if we try to inject multiple secrets, we won't mount multiple volumes.
+		// We use Memory as the storage medium for volume source to avoid
+		vol := corev1.Volume{
+			Name: AWSSecretsVolumeName,
+			VolumeSource: corev1.VolumeSource{
+				EmptyDir: &corev1.EmptyDirVolumeSource{
+					Medium: corev1.StorageMediumMemory,
+				},
+			},
+		}
+
+		p.Spec.Volumes = appendVolumeIfNotExists(p.Spec.Volumes, vol)
+		p.Spec.InitContainers = append(p.Spec.InitContainers, createAWSSidecarContainer(i.cfg, p, secret))
+
+		secretVolumeMount := corev1.VolumeMount{
+			Name:      AWSSecretsVolumeName,
+			ReadOnly:  true,
+			MountPath: filepath.Join(AWSSecretMountPathPrefix...),
+		}
+
+		p.Spec.Containers = AppendVolumeMounts(p.Spec.Containers, secretVolumeMount)
+		p.Spec.InitContainers = AppendVolumeMounts(p.Spec.InitContainers, secretVolumeMount)
+
+		// Inject AWS secret-inject webhook annotations to mount the secret in a predictable location.
+		envVars := []corev1.EnvVar{
+			// Set environment variable to let the container know where to find the mounted files.
+			{
+				Name:  SecretPathDefaultDirEnvVar,
+				Value: filepath.Join(AWSSecretMountPathPrefix...),
+			},
+			// Sets an empty prefix to let the containers know the file names will match the secret keys as-is.
+			{
+				Name:  SecretPathFilePrefixEnvVar,
+				Value: "",
+			},
+		}
+
+		for _, envVar := range envVars {
+			p.Spec.InitContainers = AppendEnvVars(p.Spec.InitContainers, envVar)
+			p.Spec.Containers = AppendEnvVars(p.Spec.Containers, envVar)
+		}
+	case core.Secret_ENV_VAR:
+		fallthrough
+	default:
+		err := fmt.Errorf("unrecognized mount requirement [%v] for secret [%v]", secret.MountRequirement.String(), secret.Key)
+		logger.Error(ctx, err)
+		return p, false, err
+	}
+
+	return p, true, nil
+}
+
+func createAWSSidecarContainer(cfg config.AWSSecretManagerConfig, p *corev1.Pod, secret *core.Secret) corev1.Container {
```

Comment:
is this a sidecar or an init container. If sidecar, isnt there a race condition, between the mounting and the user container trying to access the secret?